<a href="https://colab.research.google.com/github/dev-sayantan/MainCrafts-Technology-Data-Science-Internship/blob/main/student_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <strong>Task 01 - <u>STUDENT PERFORMANCE DATA ANALYSIS</u></strong>


---



In [11]:
# REQUIRED IMPORTS - PYTHON DATA ANALYSIS & VISUALIZATION LIBRARIES
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [12]:
# IMPORT PYTHON'S zipfile MODULE TO EXTRACT SPECIFIC FILE FROM ZIP ARCHIVE
from zipfile import ZipFile

# EXTRACTING PROCESS :-
# OPEN THE ZIP FILE
with ZipFile('/content/drive/MyDrive/Colab Notebooks/MainCrafts_Technology_DataSc_Internship_Tasks/Datasets/student.zip') as zObject:
# SPECIFIC FILE EXTRACTION(student-mat.csv) FROM ZIP ARCHIVE TO "Extracted" FOLDER
  zObject.extract('student-mat.csv', path='/content/drive/MyDrive/Colab Notebooks/MainCrafts_Technology_DataSc_Internship_Tasks/Datasets/Extracted')

In [13]:
# RUN & VIEW THE DATASET
extracted_stumat = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/MainCrafts_Technology_DataSc_Internship_Tasks/Datasets/Extracted/student-mat.csv',sep=';')
print(extracted_stumat)

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health 

In [14]:
# EXPLORING FURTHER :- CHECKING SHAPE OF THE DATAFRAME
print(f"Shape of DataFrame: {extracted_stumat.shape}")

Shape of DataFrame: (395, 33)


In [15]:
# EXPLORING FURTHER :- CHECKING DATATYPE OF THE DATAFRAME & ALSO FOR ALL FEATURES/COLUMNS
print(extracted_stumat.dtypes)

school        object
sex           object
age            int64
address       object
famsize       object
Pstatus       object
Medu           int64
Fedu           int64
Mjob          object
Fjob          object
reason        object
guardian      object
traveltime     int64
studytime      int64
failures       int64
schoolsup     object
famsup        object
paid          object
activities    object
nursery       object
higher        object
internet      object
romantic      object
famrel         int64
freetime       int64
goout          int64
Dalc           int64
Walc           int64
health         int64
absences       int64
G1             int64
G2             int64
G3             int64
dtype: object


In [16]:
# CHECKING / IDENTIFYING ALL MISSING VALUES, IF ANY, PRESENT IN THE DATAFRAME "extracted_stumat"
missing_values = extracted_stumat.isnull()
missing_value_count = extracted_stumat.isnull().sum()
print(missing_values)
print(missing_value_count)
# CHECKING THE SHAPE OF DATAFRAME OBJECT HAVING ONLY MISSING VALUES
print(missing_values.shape)

     school    sex    age  address  famsize  Pstatus   Medu   Fedu   Mjob  \
0     False  False  False    False    False    False  False  False  False   
1     False  False  False    False    False    False  False  False  False   
2     False  False  False    False    False    False  False  False  False   
3     False  False  False    False    False    False  False  False  False   
4     False  False  False    False    False    False  False  False  False   
..      ...    ...    ...      ...      ...      ...    ...    ...    ...   
390   False  False  False    False    False    False  False  False  False   
391   False  False  False    False    False    False  False  False  False   
392   False  False  False    False    False    False  False  False  False   
393   False  False  False    False    False    False  False  False  False   
394   False  False  False    False    False    False  False  False  False   

      Fjob  ...  famrel  freetime  goout   Dalc   Walc  health  absences  \

In [17]:
# REMOVING DUPLICATE VALUES PRESENT IN THE DATAFRAME
extracted_uniqueVal_stumat = extracted_stumat.drop_duplicates(keep=False)
extracted_uniqueVal_stumat

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

In [18]:
# CHECKING THE SHAPE OF THE DATAFRAME TO CONFIRM THAT THERE ARE NO DUPLICATE VALUES PRESENT IN THE DATASET
if extracted_stumat.shape == extracted_uniqueVal_stumat.shape:
  print("No Duplicate Values Found")
else:
  print("Duplicate Values Found :\n", extracted_stumat.duplicated(keep=False))

No Duplicate Values Found


In [19]:
extracted_stumat_descrStat = extracted_stumat.describe()
extracted_stumat_descrStat

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

## ANALYSIS QUESTIONS


---



#####

#### AVERAGE FINAL GRADE (G3)

In [22]:
# AVERAGE FINAL GRADE(G3) :-
# METHOD ONE -->
print("Average Final Grade(G3) [From Descriptive Statistics Table Obtained From DataFrame]:", extracted_stumat_descrStat.loc['mean', 'G3'])
# METHOD TWO -->
print("Average Final Grade(G3) [From The Cleaned DataFrame, After Performing Missing & Duplicate Values Removal]:", extracted_stumat['G3'].mean())

Average Final Grade(G3) [From Descriptive Statistics Table Obtained From DataFrame]: 10.415189873417722
Average Final Grade(G3) [From The Cleaned DataFrame, After Performing Missing & Duplicate Values Removal]: 10.415189873417722


##### Analysis Outcome --> I have calculated in two methods using two different DataFrames,
* First one obtained after generating descriptive statistics and applying pandas accessor .loc[] used for selecting data from a DataFrame by its labels (the row and column names), specifically used for label-based indexing and selection.
* The second method is about filtering a required part of the DataFrame, based on specific required feature/column name. The average final grade obtained by using the Pandas mean() function compute the arithmetic mean (average) of data within a Series or DataFrame.  

#### HOW MANY STUDENTS SCORED ABOVE 15?

In [23]:
extracted_stumat

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

In [24]:
stu_G3_Above15_totalCount = extracted_stumat[extracted_stumat['G3'] > 15].value_counts().sum()
print("Number of Students Who Scored Above 15 are:", stu_G3_Above15_totalCount)

Number of Students Who Scored Above 15 are: 40


##### Analysis Outcome --> By performing BOOLEAN INDEXING to FILTER the DATAFRAME to obtain only those students who got above 15, then by using value_counts() function, returning a SERIES per distint row, containing the frequency/occurance of students having more than 15, within the DATAFRAME & finally using sum() to aggregate the frequencies/occurances.

#### IS studytime CORRELATED WITH performance?

In [25]:
print(extracted_stumat.columns)

print(extracted_stumat.corr(numeric_only=True))

print("Correlation Coefficient value, between two features/columns: 'studytime' & 'G3' is =", extracted_stumat['studytime'].corr(extracted_stumat['G3']))

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')
                 age      Medu      Fedu  traveltime  studytime  failures  \
age         1.000000 -0.163658 -0.163438    0.070641  -0.004140  0.243665   
Medu       -0.163658  1.000000  0.623455   -0.171639   0.064944 -0.236680   
Fedu       -0.163438  0.623455  1.000000   -0.158194  -0.009175 -0.250408   
traveltime  0.070641 -0.171639 -0.158194    1.000000  -0.100909  0.092239   
studytime  -0.004140  0.064944 -0.009175   -0.100909   1.000000 -0.173563   
failures    0.243665 -0.236680 -0.250408    0.092239  -0.173563  1.000000   
famrel      0.053940 -0.003914 -0.001370   -0.016808   0.039731 -0.044337   

##### Analysis Outcome --> After checking all the column names, printing correlation matrix having correlation coefficient values between features with only numeric datatypes, ignoring the non-numeric datatypes. Then, checking correlation between the two columns only - 'studytime' & 'G3', indicating that it has a very low/weak positive correlation between them, 0.0978.

#### WHICH GENDER PERFORMS BETTER ON AVERAGE?

In [29]:
gender_count = extracted_stumat['sex'].value_counts()
print(gender_count)
extracted_stumat['G3'].groupby(extracted_stumat['sex']).mean()

sex
F    208
M    187
Name: count, dtype: int64


sex
F     9.966346
M    10.914439
Name: G3, dtype: float64

##### Analysis Outcome --> So, above output is clearly showing that 187 Male students are having average final grade(G3) of 10.9, and thus performs better than 208 Female students who secured 9.96 as average final grade(G3).